src/v/serde/serde.h

```diff
@@ -171,53 +170,69 @@ void write(iobuf& out, T t) {
+std::decay_t<T> read(iobuf_parser& in) {
+    return read<T>(in, 0U);
+}
+
+template<typename T>
+header read_header(iobuf_parser& in, std::size_t const bytes_left_limit) {
     using Type = std::decay_t<T>;
 
-    auto const version = read<version_t>(in);
-    auto const compat_version = read<version_t>(in);
-    auto const size = read<serde_size_t>(in);
+    auto const version = read<version_t>(in, bytes_left_limit);
+    auto const compat_version = read<version_t>(in, bytes_left_limit);
+    auto const size = read<serde_size_t>(in, bytes_left_limit);
 
     if (unlikely(compat_version > Type::redpanda_serde_version)) {
         throw serde_exception(fmt_with_ctx(
           ssx::sformat,
-          "read compat_version={} > {}::version={}",
+          "read {}: compat_version={} > {}::version={}",
+          type_str<Type>(),
           static_cast<int>(compat_version),
           type_str<T>(),
           static_cast<int>(Type::redpanda_serde_version)));
     }
 
-    if (unlikely(in.bytes_left() < size)) {
-        throw serde_exception(fmt_with_ctx(
-          ssx::sformat,
-          "bytes_left={}, size={}",
-          in.bytes_left(),
-          static_cast<int>(size)));
-    }
-
-    return std::make_tuple(version, compat_version, size);
+    return header{
+      ._version = version,
+      ._compat_version = compat_version,
+      ._bytes_left_limit = in.bytes_left() - size};
 }
 
 template<typename T>
-std::decay_t<T> read(iobuf_parser& in) {
+std::decay_t<T> read(iobuf_parser& in, std::size_t const bytes_left_limit) {
     using Type = std::decay_t<T>;
     static_assert(has_serde_read<T> || is_serde_compatible_v<Type>);
 
     auto t = Type();
     if constexpr (is_envelope_v<Type>) {
-        [[maybe_unused]] auto const [version, compat_version, size]
-          = read_header<Type>(in);
+        auto const h = read_header<Type>(in, bytes_left_limit);
         if constexpr (has_serde_read<Type>) {
-            t.serde_read(in, version, compat_version, size);
+            t.serde_read(in, h);
         } else {
-            envelope_for_each_field(
-              t, [&](auto& f) { f = read<std::decay_t<decltype(f)>>(in); });
+            envelope_for_each_field(t, [&](auto& f) {
+                using FieldType = std::decay_t<decltype(f)>;
+                if (h._bytes_left_limit == in.bytes_left()) {
+                    return false;
+                }
+                if (unlikely(in.bytes_left() < h._bytes_left_limit)) {
+                    throw serde_exception(fmt_with_ctx(
+                      ssx::sformat,
+                      "field spill over in {}, field type {}: envelope_end={}, "
+                      "in.bytes_left()={}",
+                      type_str<Type>(),
+                      type_str<FieldType>(),
+                      h._bytes_left_limit,
+                      in.bytes_left()));
+                }
+                f = read<FieldType>(in);
+                return true;
+            });
```

Comment:
Test for skip data not read. If we have less fields, then we should skip all data that is coming after the fields we have initialized.